### PR TITLE
Fixed XR shader failure in URP ShaderGraph

### PR DESCRIPTION
**Pull request by thomas-zeng**
### Checklist for PR maker
---
### Purpose of this PR
- https://fogbugz.unity3d.com/f/cases/1185272/
- https://fogbugz.unity3d.com/f/cases/1188487/
- https://fogbugz.unity3d.com/f/cases/1195207/
---
### Testing status

**Manual Tests**: What did you do?
- [x] Opened test project + Run graphic tests locally
- [x] Built a player
- [ ] Checked new UI names with UX convention
- [ ] Tested UI multi-edition + Undo/Redo
- [x] C# and shader warnings (supress shader cache to see them)
- [ ] Checked new resources path for the reloader (in devloper mode, you have a button at end of resources that check the pathes)
- Other: 

**Automated Tests**: What did you setup? (Add a screenshot or the reference image of the test please)

**Yamato**: (Select your branch):
https://yamato.prd.cds.internal.unity3d.com/jobs/78-ScriptableRenderPipeline/tree/urp%252Fxr%252Ffix-sg-compile-err

Any test projects to go with this to help reviewers?

---
### Comments to reviewers
- HDRP might have similar issues with Shader Graph. HDSubShaderUtilities.cs will require similar changes to be XR aware.
- QA: please save Shader graph asset and clear shader cache before testing.

### CODE REVIEW

File: com.unity.render-pipelines.core/ShaderLibrary/UnityInstancing.hlsl [hunk 1]
```diff
@@ -96,14 +96,17 @@
 // - UNITY_SETUP_STEREO_EYE_INDEX_POST_VERTEX
 #ifdef UNITY_STEREO_INSTANCING_ENABLED
 #if defined(SHADER_API_GLES3) || defined(SHADER_API_GLCORE)
-    #define DEFAULT_UNITY_VERTEX_OUTPUT_STEREO                          uint stereoTargetEyeIndexSV : SV_RenderTargetArrayIndex; uint stereoTargetEyeIndex : BLENDINDICES0;
-    #define DEFAULT_UNITY_INITIALIZE_VERTEX_OUTPUT_STEREO(output)       output.stereoTargetEyeIndexSV = unity_StereoEyeIndex; output.stereoTargetEyeIndex = unity_StereoEyeIndex;
+    #define DEFAULT_UNITY_VERTEX_OUTPUT_STEREO                          uint stereoTargetEyeIndexSV : SV_RenderTargetArrayIndex; uint stereoTargetEyeIndexGL : BLENDINDICES0;
+    #define DEFAULT_UNITY_INITIALIZE_VERTEX_OUTPUT_STEREO(output)       output.stereoTargetEyeIndexSV = unity_StereoEyeIndex; output.stereoTargetEyeIndexGL = unity_StereoEyeIndex;
+    #define DEFAULT_UNITY_TRANSFER_VERTEX_OUTPUT_STEREO(input, output)  output.stereoTargetEyeIndexGL = input.stereoTargetEyeIndexGL;
```
Comment (fabien-unity) [inline com.unity.render-pipelines.core/ShaderLibrary/UnityInstancing.hlsl]: Where is stereoTargetEyeIndexGL defined ? I can't find it.

File: com.unity.render-pipelines.universal/Editor/ShaderGraph/UniversalShaderGraphResources.cs [hunk 2]
```diff
@@ -85,6 +85,12 @@ internal struct Varyings
             uint instanceID;
             [Semantic("FRONT_FACE_SEMANTIC")][SystemGenerated][OverrideType("FRONT_FACE_TYPE")][PreprocessorIf("defined(SHADER_STAGE_FRAGMENT) && defined(VARYINGS_NEED_CULLFACE)")]
             bool cullFace;
+            [Semantic("SV_RenderTargetArrayIndex")] [PreprocessorIf("(defined(UNITY_STEREO_INSTANCING_ENABLED) && (defined(SHADER_API_GLES3) || defined(SHADER_API_GLCORE)))")]
+            uint stereoTargetEyeIndexSV;
+            [Semantic("BLENDINDICES0")] [PreprocessorIf("(defined(UNITY_STEREO_INSTANCING_ENABLED) && !(defined(SHADER_API_GLES3) || defined(SHADER_API_GLCORE)))")]
```
Comment (fabien-unity) [inline com.unity.render-pipelines.universal/Editor/ShaderGraph/UniversalShaderGraphResources.cs]: Why is the test negated here ?
It should be the same condition for stereoTargetEyeIndexSV and stereoTargetEyeIndexGL, no ?

Comment (thomas-zeng) [reply com.unity.render-pipelines.universal/Editor/ShaderGraph/UniversalShaderGraphResources.cs]: You are right, fixed in 
4884a9d

**Comment by thomas-zeng** [comment]
@Kink3d Hey Matt, will you be the right person to review URP-SG changes? let me know if I should add other SG dev as reviewers :)

**Comment by thomas-zeng** [comment]
@ROBYER1 
Hi!
I tried that water shader in my project and I cannot repro your case. which version of unity you are using when testing?

**Comment by ROBYER1** [comment]
@thomas-zeng I'm glad you found one of my fogbugz submissions, if you need any other more recent versions of that test project let me know. I am using it with the current 2019.3 beta releases currently at beta 9 of 2019.3. 

Related cases on fogbugz: 1188487 , 1195207 , 1190116

What I did notice was that sometimes the StereoEye issues would not show until I tried to build an apk - I encourage you to try building an APK as it would often fail there. Once built and running on an Oculus Quest or Go, that water material and the paint shader graph from the Sample Scene (Paint 1G or whatever it is called) won't render properly.

Thankyou so much for looking into this, your hard work is appreciated

**Comment by thomas-zeng** [comment]
@ROBYER1 please try 28857bf fix!

**Comment by ROBYER1** [comment]
@thomas-zeng giving it a shot now, would be interested to see how the 28857bf fixes the SG stereoeye issue as the test project is for Single-Pass :O I'll report back.

Will be so pleased if that test project finally builds, wish I commented on these commits earlier on been wrestling with QA to show them what was going on but I understand they have a job to do to identify issues.

**Comment by ROBYER1** [comment]
@thomas-zeng  

So it successfully built, however the water material does not render, the right eye is warped and not aligned so your view is cross-eyed (similar to the experience when Post processing was broken for VR on Oculus Quest/Go) and the app shortly crashes, in logcat I get the errors listed below and see my screenshot attached.

This is already reported on Fogbugz at 1195463

Error Unity: OPENGL NATIVE PLUG-IN ERROR: GL_INVALID_OPERATION: Operation illegal in current state
Error Unity: (Filename: ./Runtime/GfxDevice/opengles/GfxDeviceGLES.cpp Line: 353)

Can you confirm the same happens for you when you build and run on an Oculus Quest or Go? It makes it unusable to develop with if these issues only crop up at runtime after building with no errors.
native plugi n error


Interestingly, the speedtree I have in the scene still has very dark artefacts on the leaves which look very incorrect compared to its behaviour in the 3d built-in renderer which I reported months ago at 1185340 if I can pass that onto anyone working on that area?

**Comment by ROBYER1** [comment]
@thomas-zeng Did you test the scene on the device at all from my post above? I would like to know if you see the same OPENGL Native errors on the Oculus Quest. If not I would be happy to provide an example project as per fogbugz case 1195463

@ryanhy-unity is there any way we can test the Shader Graph features on Android with instancing to prevent these issues from arising? I'm trying my best to highlight these issues through QA and fogbugz but the best solution would be to automate the testing of these features on Android where possible to make sure the issues are flagged. It doesn't sound great to hear that the issues might be fixed in another PR that doesn't exist yet.

On another note, improvements to integration with stereo rendering are welcomed and your hard work is appreciated!

**Comment by thomas-zeng** [comment]
@ROBYER1 Hi Rob, Thank you for reporting back your findings. I let our QA tested your water shader and your scene and we are aware that Oculus Quest has issues regarding shader graph shaders. 

My main expertise is not on Quest so what I did is I forwarded Quest specific issues to my colleague. 
This PR fixes XR Shader Graph compiling issues for all platforms which unblock the Quest team's further investigation. That is the main reason why we are fixing your concerns in multiple PRs.

**Comment by thomas-zeng** [comment]
@phi-lira Hey felipe! This is a high priority PR, could you review this on behalf of the URP team when you have time?

**Comment by alindmanUnity** [comment]
adding GFX qa bellevue to verify shader graph related changes

**Comment by LandonTownsendUnity** [comment]
Approved again

**Comment by ROBYER1** [comment]
> Smoke tests passed:
> PBR master node (all features)
> Unlit master node (all features)
> All nodes connected to PBR color (error test)
> All nodes connected to Unlit color (error test)

Did they pass for Android VR? Me and @soraryu have reported a few issues related to this when running builds on the Oculus Quest, sometimes things work fine in editor but in the build they do not (see my comments further up) and see Case 1195463

File: com.unity.render-pipelines.core/ShaderLibrary/UnityInstancing.hlsl [hunk 3]
```diff
@@ -96,14 +96,17 @@
...
     #define DEFAULT_UNITY_VERTEX_OUTPUT_STEREO                          uint stereoTargetEyeIndex : SV_RenderTargetArrayIndex;
     #define DEFAULT_UNITY_INITIALIZE_VERTEX_OUTPUT_STEREO(output)       output.stereoTargetEyeIndex = unity_StereoEyeIndex
-#endif
     #define DEFAULT_UNITY_TRANSFER_VERTEX_OUTPUT_STEREO(input, output)  output.stereoTargetEyeIndex = input.stereoTargetEyeIndex;
-    #define DEFAULT_UNITY_SETUP_STEREO_EYE_INDEX_POST_VERTEX(input) unity_StereoEyeIndex = input.stereoTargetEyeIndex;
+    #define DEFAULT_UNITY_SETUP_STEREO_EYE_INDEX_POST_VERTEX(input)     unity_StereoEyeIndex = input.stereoTargetEyeIndex;
+#endif
+
 #elif defined(UNITY_STEREO_MULTIVIEW_ENABLED)
     #define DEFAULT_UNITY_VERTEX_OUTPUT_STEREO float stereoTargetEyeIndex : BLENDWEIGHT0;
```
Comment (phi-lira) [inline com.unity.render-pipelines.core/ShaderLibrary/UnityInstancing.hlsl]: Shouldn't this be stereoTargetEyeIndexGL as well? Seems like you renamed the ones above but not this.

**Comment by phi-lira** [comment]
I agree with @Verasl it would be great to land Stereo Shader Graph tests together with this.

**Comment by soraryu** [comment]
I'll merge this one here into my earlier pull request, https://github.com/Unity-Technologies/ScriptableRenderPipeline/pull/4802.

**Comment by phi-lira** [comment]
@thomas-zeng to confirm understanding. The issues reported by 
@soraryu and @ROBYER1 in Quest are going to be fixed in a follow up PR, is that correct?

File: com.unity.render-pipelines.core/ShaderLibrary/UnityInstancing.hlsl (hunk 3 above)
Comment (thomas-zeng) [reply com.unity.render-pipelines.core/ShaderLibrary/UnityInstancing.hlsl]: This is by design. stereoTargetEyeIndexGL is only for GLCore and GL3.

**Comment by thomas-zeng** [comment]
@Verasl I completely agree. Who should I talk to regarding shader graph auto tests? 
Since shader graph is generating shaders, it is important shader graph devs write tests around XR along with the development process.

**Comment by ROBYER1** [comment]
@thomas-zeng @Verasl Who is setting up Shader Graph tests for XR?

I would like to make sure that Mobile/Android VR for Oculus Quest/Go isn't overlooked there as the reproducible issues for those platforms reported by me and @soraryu aren't being caught in tests for PRs like this. 

I would be happy to assist with any feedback and reporting as Shader graph usually looks fine in the editor but when built and running on the Oculus Quest the issues only show up then

See Fogbugz Case 1195463

**Comment by phi-lira** [comment]
I'm following up with team about getting test in a follow up PR. We want to get this PR with fixes on a 7.1.4 package. 

@ROBYER1 @soraryu to confirm. What fb cases that impact you and are not resolved by this PR? I captured the following ones so I can follow up.
- 1190116
- 1195463
- 1188487

**Comment by ROBYER1** [comment]
> I'm following up with team about getting test in a follow up PR. We want to get this PR with fixes on a 7.1.4 package.
> 
> @ROBYER1 @soraryu to confirm. What fb cases that impact you and are not resolved by this PR? I captured the following ones so I can follow up.
> 
> * 1190116
> * 1195463

@phi-lira thankyou for chasing this, I have:

- 1188127 - UI not rendering in builds on Quest and possibly linked to the issue below (shown in screenshot in previous comments on this PR)
- 1195463 - You already have this one
- 1185340 - Shader graph Speedtree nodes look incorrect and default generated Shader Graph URP Speedtrees exhibit the same issue.

With regards to 1195463 a test of core Shader graph functionality for Android VR would be brilliant otherwise as we continue to use Shader graph on mobile VR platforms I may find more OpenGL native  plug-in errors at runtime in builds which aren't caught by the Editor.

**Comment by soraryu** [comment]
@phi-lira 

- 1190415 | [URP] Light glitches related to SRP Batcher
- 1185839 | [URP] Enabling GPU Instancing on ShaderGraph-based material breaks Single Pass stereo rendering
- 1185837 | [URP] Single Pass Instanced VR rendering breaks ShaderGraph shaders

Note that all of these have been closed as duplicates and I neither know nor can access the "original" bugs.

**Comment by phi-lira** [comment]
@soraryu 
1190415 if fixed in Unity side and backported to 2019.3 duplicate of https://issuetracker.unity3d.com/product/unity/issues/guid/1179976
1185839 is not qualified meaning it's either not reproducible or there's not enough info.
1185837 I can't see what is this duplicated of but juding by description it's one of the issues mentioned in this PR.

**Comment by phi-lira** [comment]
URP and ShaderGraph team are coordinating to get the ShaderGraph testing coverage. It will be done in follow up PRs.

**Comment by ROBYER1** [comment]
> @thomas-zeng
> 
> So it successfully built, however the **water material** and **UI world space canvas** does not render, the right eye is warped and not aligned so your view is cross-eyed (similar to the experience when Post processing was broken for VR on Oculus Quest/Go) and the app shortly crashes, in logcat I get the errors listed below and see my screenshot attached.
> 
> This is already reported on Fogbugz at 1195463
> 
> Error Unity: OPENGL NATIVE PLUG-IN ERROR: GL_INVALID_OPERATION: Operation illegal in current state
> Error Unity: (Filename: ./Runtime/GfxDevice/opengles/GfxDeviceGLES.cpp Line: 353)
> 
> Can you confirm the same happens for you when you build and run on an Oculus Quest or Go? It makes it unusable to develop with if these issues only crop up at runtime after building with no errors.
> 
> The main reason I am testing this material is that at some point on LWRP it rendered completely fine on the device so there is a regression here somewhere in the move to URP and the changes along with it to SG/SRP
> native plugi n error
> 
> Interestingly, the speedtree I have in the scene still has very dark artefacts on the leaves which look very incorrect compared to its behaviour in the 3d built-in renderer which I reported months ago at 1185340 if I can pass that onto anyone working on that area?

@phi-lira since the merge of this into Master, the rendering of both eyes is now correct, however the Water Shader Graph + World Space Canvas UI do not render and  "OPENGL Native" errors still show in logcat until the build crashes within a few seconds.

I'm sure the shader graph team will notice this too with the test scene in Case 1195463 but it is great to see some improvement on this front from this PR merge.

I will continue testing things when I have time but hopefully that case should be a good start for the testing.

**Comment by luis-paolino** [comment]
Having the same problem with my 2 shaders made in Shader Graph.
undeclared identifier 'unity_StereoEyeIndex' 

Im using Unity 2019.3.0b8, URP/SG 7.1.2 and SRP from this git master brunch
I can't even open the APP in Oculus Quest

**Comment by soraryu** [comment]
@phi-lira @thomas-zeng there's still issues on latest master and ShaderGraph on Quest if Texture2DArray nodes are used in a ShaderGraph shader (still the same right eye offset + errors logged from Shader Compiler). Other ShaderGraph shaders seem to be fine.

**Comment by thomas-zeng** [comment]
@soraryu Thank you for reporting back! Maybe I get the fogbugz# for this?

**Comment by soraryu** [comment]
Haven't reported a bug yet (I've been repeatedly told by QA to not report bugs on unreleased versions). I'll see if I can properly repro it today and will post the case number here.

**Comment by soraryu** [comment]
That's the issue:
image

My assumption is that it is related to wrong vertex/fragment scope, but I can't fully pinpoint whether it's because of Texture2DArray or IsFrontFacing.

(Case 1201706) [ShaderGraph] Compilation Errors on shader that uses IsFrontFacing and Texture Array
Also, #5219 already has everything set up for testing (ref images are missing).

**Comment by LandonTownsendUnity** [comment]
Backported in this PR:
https://github.com/Unity-Technologies/ScriptableRenderPipeline/pull/5012

**Pull request merged**